sdks/go/pkg/beam/core/runtime/harness/harness.go

```diff
@@ -100,34 +103,39 @@ func Main(ctx context.Context, loggingEndpoint, controlEndpoint string) error {
 			return fmt.Errorf("recv failed: %v", err)
 		}
 
-		log.Debugf(ctx, "RECV: %v", proto.MarshalTextString(req))
-		recordInstructionRequest(req)
+		// Launch a goroutine to handle the control message.
+		// TODO(wcn): maybe only do this for 'heavy' messages?
+		// TODO(wcn): implement a rate limiter for 'heavy' messages?
+		go func() {
+			log.Debugf(ctx, "RECV: %v", proto.MarshalTextString(req))
+			recordInstructionRequest(req)
 
-		if isEnabled("cpu_profiling") {
-			cpuProfBuf.Reset()
-			pprof.StartCPUProfile(&cpuProfBuf)
-		}
-		resp := ctrl.handleInstruction(ctx, req)
+			if isEnabled("cpu_profiling") {
+				cpuProfBuf.Reset()
+				pprof.StartCPUProfile(&cpuProfBuf)
+			}
+			resp := ctrl.handleInstruction(ctx, req)
 
-		if isEnabled("cpu_profiling") {
-			pprof.StopCPUProfile()
-			if err := ioutil.WriteFile(fmt.Sprintf("%s/cpu_prof%s", storagePath, req.InstructionId), cpuProfBuf.Bytes(), 0644); err != nil {
-				log.Warnf(ctx, "Failed to write CPU profile for instruction %s: %v", req.InstructionId, err)
+			if isEnabled("cpu_profiling") {
+				pprof.StopCPUProfile()
+				if err := ioutil.WriteFile(fmt.Sprintf("%s/cpu_prof%s", storagePath, req.InstructionId), cpuProfBuf.Bytes(), 0644); err != nil {
+					log.Warnf(ctx, "Failed to write CPU profile for instruction %s: %v", req.InstructionId, err)
+				}
 			}
-		}
 
-		recordInstructionResponse(resp)
-		if resp != nil {
-			respc <- resp
-		}
+			recordInstructionResponse(resp)
+			if resp != nil {
+				respc <- resp
+			}
+		}()
 	}
 }
 
 type control struct {
-	plans map[string]*exec.Plan
-	data  *DataManager
-
-	// TODO: running pipelines
+	plans  map[string]*exec.Plan // protected by mu
+	active map[string]*exec.Plan // protected by mu
+	mu     sync.Mutex
+	data   *DataManager
```

Comment:
<!--new_thread; commit:e1da5f2a09fbe4caf5901a1b3d000234fd49582d; resolved:0-->
Consider adding a newline between the mutex protected group and the data field. The visual separation will reinforce the intention comments.

Reply:
Yes, good point.